lexers/rest.lua

```diff
@@ -131,122 +133,109 @@ local indented_block = function(input, index)
-  footnote_block = l.STYLE_LABEL,
-  citation_block = l.STYLE_LABEL,
-  link_block = l.STYLE_LABEL,
-  directive = l.STYLE_KEYWORD,
-  sphinx_directive = l.STYLE_KEYWORD .. ',bold',
-  unknown_directive = l.STYLE_KEYWORD .. ',italics',
-  code_block = l.STYLE_EMBEDDED .. ',eolfilled',
-  substitution = l.STYLE_VARIABLE,
-  strong = 'bold',
-  em = 'italics',
-  role = l.STYLE_CLASS,
-  interpreted = l.STYLE_STRING,
-  inline_literal = l.STYLE_EMBEDDED,
-  link = 'underlined',
-  footnote = 'underlined',
-  citation = 'underlined',
-}
+lex:add_rule('non_space', token(lexer.DEFAULT,
+                          lexer.alnum * (lexer.any - lexer.space)^0))
+lex:add_rule('escape', token(lexer.DEFAULT, '\\' * lexer.any))
+
+lex:add_style('list', lexer.STYLE_TYPE)
+lex:add_style('literal_block', lexer.STYLE_EMBEDDED .. ',eolfilled')
+lex:add_style('footnote_block', lexer.STYLE_LABEL)
+lex:add_style('citation_block', lexer.STYLE_LABEL)
+lex:add_style('link_block', lexer.STYLE_LABEL)
+lex:add_style('directive', lexer.STYLE_KEYWORD)
+lex:add_style('sphinx_directive', lexer.STYLE_KEYWORD .. ',bold')
+lex:add_style('unknown_directive', lexer.STYLE_KEYWORD .. ',italics')
+lex:add_style('code_block', lexer.STYLE_EMBEDDED .. ',eolfilled')
+lex:add_style('substitution', lexer.STYLE_VARIABLE)
+lex:add_style('strong', 'bold')
+lex:add_style('em', 'italics')
+lex:add_style('role', lexer.STYLE_CLASS)
+lex:add_style('interpreted', lexer.STYLE_STRING)
+lex:add_style('inline_literal', lexer.STYLE_EMBEDDED)
+lex:add_style('link', 'underlined')
+lex:add_style('footnote', 'underlined')
+lex:add_style('citation', 'underlined')
 
 local sphinx_levels = {
   ['#'] = 0, ['*'] = 1, ['='] = 2, ['-'] = 3, ['^'] = 4, ['"'] = 5
 }
 
--- Section-based folding.
-M._fold = function(text, start_pos, start_line, start_level)
-  local folds, line_starts = {}, {}
-  for pos in (text .. '\n'):gmatch('().-\r?\n') do
-    line_starts[#line_starts + 1] = pos
-  end
-  local style_at, CONSTANT, level = l.style_at, l.CONSTANT, start_level
-  local sphinx = l.property_int['fold.by.sphinx.convention'] > 0
-  local FOLD_BASE = l.FOLD_BASE
-  local FOLD_HEADER, FOLD_BLANK = l.FOLD_HEADER, l.FOLD_BLANK
-  for i = 1, #line_starts do
-    local pos, next_pos = line_starts[i], line_starts[i + 1]
-    local c = text:sub(pos, pos)
-    local line_num = start_line + i - 1
-    folds[line_num] = level
-    if style_at[start_pos + pos - 1] == CONSTANT and c:find('^[^%w%s]') then
-      local sphinx_level = FOLD_BASE + (sphinx_levels[c] or #sphinx_levels)
-      level = not sphinx and level - 1 or sphinx_level
-      if level < FOLD_BASE then level = FOLD_BASE end
-      folds[line_num - 1], folds[line_num] = level, level + FOLD_HEADER
-      level = (not sphinx and level or sphinx_level) + 1
-    elseif c == '\r' or c == '\n' then
-      folds[line_num] = level + FOLD_BLANK
-    end
-  end
-  return folds
-end
-
-l.property['fold.by.sphinx.convention'] = '0'
+-- -- Section-based folding.
```

Comment:
This function should be probably replaced by the series of `lexer.add_fold_point()` calls, but I have no idea how.

@orbitalquark Would you be able to help, please?

Reply:
You should be able to just change the name from `_fold` to `fold`. This will override the default `lexer.fold()` implementation.

Reply:
`lex.fold()`. If you change `lexer.fold()` then that affects all lexers, which would be bad.

Reply:
And what about that
```lua
l.property['fold.by.sphinx.convention'] = '0'
```
`lex.property` is apparently `nil`, so my editor breaks on this line. I have commented it out for now.

Reply:
Did you try `lexer.property[]`? That field belongs to the `lexer` module, not a lexer object.

Reply:
Thank you.